### ISSUE TITLE
super long error messages are hard to read

### ISSUE THREAD

**Issue by devoncarew** [dart-lang/dev_compiler#322]
Not sure what the resolution for this is, but the error messages are very long, and hard to read:

```
severe: [StaticTypeError] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<V> (package:collection/equality.dart, line 264, col 44)
severe: [InvalidMethodOverride] Invalid override. The type of UnmodifiableMapMixin.remove ((K) → V) is not a subtype of Map<K, V>.remove ((Object) → V). (package:collection/src/unmodifiable_wrappers.dart, line 230, col 3)
severe: [InvalidMethodOverride] Invalid override. The type of DelegatingSet.lookup ((E) → E) is not a subtype of Set<E>.lookup ((Object) → E). (package:collection/wrappers.dart, line 237, col 3)
```

**Comment by devoncarew** [comment]
Perhaps:

```
severe:
  [StaticTypeError] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<V>, at package:collection/equality.dart line 264
  [InvalidMethodOverride] Invalid override. The type of UnmodifiableMapMixin.remove ((K) → V) is not a subtype of Map<K, V>.remove ((Object) → V), at package:collection/src/unmodifiable_wrappers.dart line 230
  [InvalidMethodOverride] Invalid override. The type of DelegatingSet.lookup ((E) → E) is not a subtype of Set<E>.lookup ((Object) → E), at package:collection/wrappers.dart line 237
```

**Comment by devoncarew** [comment]
Though, that's not really much shorter...

**Comment by vsmenon** [comment]
Perhaps we should use something closer to the analyzer's wording.

"A value of type 'DefaultEquality' cannot be assigned to a variable of type 'Equality<V>'"

"The parameter type 'K' is not a supertype of 'Object' as required by the method it is overriding in class Map"

@leafpetersen - will we basically get this for free with your type system integration?

**Comment by leafpetersen** [comment]
For the moment at least our integrated type system simply uses the analyzer's error messages, so yes.

**Comment by devoncarew** [comment]
Here's another redux:

```
Compiling web/foo_bar.dart...

errors:

package:collection/equality.dart, line 264, col 44
  [StaticTypeError] Type check failed: const DefaultEquality() (DefaultEquality) is not of type Equality<V>
package:collection/src/unmodifiable_wrappers.dart, line 230, col 3
  [InvalidMethodOverride] Invalid override. The type of UnmodifiableMapMixin.remove ((K) → V) is not a subtype of Map<K, V>.remove ((Object) → V).
package:collection/wrappers.dart, line 237, col 3
  [InvalidMethodOverride] Invalid override. The type of DelegatingSet.lookup ((E) → E) is not a subtype of Set<E>.lookup ((Object) → E).

3 errors found; wrote output to web/foo_bar.dart.js.
```

It:
- tells people what is being compiled
- gives a summary at the end of the issues found and where the output is
- breaks the issues into categories (error/warning/info) and adds whitespace to make them easy to parse into categories
- gives nearly a whole line to the error message, which can be pretty long
- helps people scan for which packages are producing which errors

Thoughts?

**Comment by bwilkerson** [comment]
I don't have a strong preference for the format, but whatever error format we have, I'd really like to have it be consistent between the command-line analyzer and ddc.

**Comment by jmesserly** [comment]
+1, I think the right approach here is to put the relevant logic into analyzer_cli. That way we get nice errors everywhere :)

**Comment by devoncarew** [comment]
I think the analyzer_cli benefits from terseness and being able to emit one error per line. That's a little difficult with the ddc error messages. Unless the ddc error messages get much shorter, some info will probably need to be split across lines. Plus, it's helpful to be able to consume the ddc output with  errors bundled up by packages.

**Comment by jmesserly** [comment]
@devoncarew -- I don't understand your comment. we are integrating strong mode into Analyzer. There will be no such thing as "ddc error messages"

**Comment by devoncarew** [comment]
I assume you'll see output when you run `dev_compiler` on your app to compile it - that it would should you all the errors / warnings / infos that would affect your compiled output, both from the current package, and from all referenced packages.

**Comment by jmesserly** [comment]
that's the same as analyzer_cli when you pass --show-package-warnings right?

**Comment by devoncarew** [comment]
Well, I don't know :) Would it be? I suppose there might be SDK issues as well.

In any case, my feedback here is that the current format of the ddc errors - combined with the message length - means that the output of ddc is really hard to parse. Happy to help noodle about ways to improve it if you think it's important -

**Comment by jmesserly** [comment]
Totally agree on the problem :). I'm not communicating this well...

Basically, DDC errors == bad. So... we're removing them!

In other words: we're fixing this by integrating with Analyzer, reusing Analyzer messages, formats, etc.

More shared code === less people solving the same problems :). If we improve error message UX in one place, everyone will benefit.

Let me know if this sounds good to you?

**Comment by devoncarew** [comment]
I'm not sure the use cases are the same - I care about more things, and differently, when compiling an app to JS vs. when analyzing a library or project.

It might be best to talk in person, and would be good to know what the errors look like after analyzer integration.

I took a stab at creating a compilation report file, which presents the information differently than the CLI output. You definitely have a lot of limitations for presentation options w/ a CLI and the large volume of info doesn't help either.

Here's a sample output file: http://dart-atom.github.io/dartlang/entry_dart_results.html.

**Comment by devoncarew** [comment]
Closing in favor of https://github.com/dart-lang/dev_compiler/issues/377.
